Ieee1609.2.1

wooks-ssoltech opened the pull request:
reopen pull request for adding TCI message for IEEE1609.2.1 testing.

wooks-ssoltech committed:
IMPORT CompactTimeConfidence missing

wooks-ssoltech committed:
IMPORT CompactTimeConfidence missing

wooks-ssoltech committed:
added TCIMessage for IEEE1609.2.1

wooks-ssoltech committed:
add TCI160921.asn
add TCI160921.asn to the script Create_TCall.bat
add Indication for IEEE1609.2.1

neil-keysight-com commented on ASN1/TCIindication.asn:
To allow for further future extension of the Event enumeration without breaking encoding compatibility, we should avoid using as many explicit values for the dot2dot1 events. I suggest changing the Event entry to a single ‘eDot2Dot1’ value. We can then extend the EventParams choice to include a new entry for dot2dot1Params. As an example:


EventParams ::= CHOICE{
    service         [0]     ServiceParameters,
    wsm             [1]     WsmParameters,
    ip              [2]     IpParameters,
    radioframe      [3]     RadioParameters,
    security        [4]     SecResultParams,
    ...
   dot2dot1 [5] Dot2Dot1Params
}

Dot2Dot1Params ::= SEQUENCE {
   state Dot2Dot1State,
   …
}

Dot2Dot1State ::= ENUMERATED {
   eInitialised (1),
   eEnrolled (2),
   eAuthReqResp (3),
   eAuthDownResp (4),
   eProvisioned (5),
   exception (15) --exception added with large value to allow for future expansion
}

wooks-ssoltech replied:
Ok  I an going to revise the code and request merge. Thanks

wooks-ssoltech committed:
TCI160921 Frame added

wooks-ssoltech committed:
revised as Neil's review.
only one eDot2Dot1 event used instead of adding all dot2dot1 state change event as events.
extend the EventParams choice to include a new entry for dot2dot1Params.

wooks-ssoltech committed:
Dot2Dot1Indication event, eventParams changed

neil-keysight-com reviewed (approved):
Changes appear backwards compatible per compilation and manual examination. Should be good to merge to main branch.

tetris02 committed:
add new tci

tetris02 committed:
d160921 Frame 위치 변경

wooks-ssoltech committed:
Merge branch 'IEEE1609.2.1' of github.com:ssoltech/OmniAir-Test-Control-Interface into IEEE1609.2.1

cv2x reviewed (changes requested):
@wooks-ssoltech   I think you accidentally checked in a merge conflict .. 
I checked out the pull request, and I see the "<<<<<< Head" in the actual file, as is shown here.
Also  , please run in "dos2unix" to get rid of the ^M dos newlines.

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -1,3 +1,156 @@
+<<<<<<< HEAD
```
you can see the accidentally checked-in merge conflict markers here.

wooks-ssoltech replied:
fixed in [1888a7c](https://github.com/OATCI/OmniAir-Test-Control-Interface/pull/8/commits/1888a7c53db93e6b4a5bf8281876574cf227f147)

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -145,4 +315,5 @@ SecResultParams ::= SEQUENCE { -- used in EventParams
-END
\ No newline at end of file
+END
+>>>>>>> 4f5e317007a18d1d2e2ee53f7c5fbac06856ea12
```
similar issue here.

wooks-ssoltech replied:
I fixed it in commit [1888a7c](https://github.com/OATCI/OmniAir-Test-Control-Interface/pull/8/commits/1888a7c53db93e6b4a5bf8281876574cf227f147)

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -106,6 +261,21 @@ SecResultParams ::= SEQUENCE { -- used in EventParams
+    eAuthReqResp (3),
+    eAuthDownResp (4),
+    eProvisioned (5),
+    exception (15) --exception added with large value to allow for future expansion
```
should this have an extension marker?

wooks-ssoltech replied:
I just follow the convention that other params have an extension marker

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -49,6 +202,7 @@ Event ::= ENUMERATED {
     eDot3RequestMatchedAvailAppService  (8),    -- request matched with available application-service as per WME-Notification.indication
     eDot2VerificationCompleteWithResult (9),    -- message signature verification is complete
+    eDot2Dot1                           (10),   -- dot2dot1 event when PKI state is changed
     exception                           (15)    -- exception reported from IUT
```
extension marker recommended.

cv2x reviewed (changes requested):
see comments.

wooks-ssoltech committed:
1. add TCI Messages
- DownloadCcf
- DownloadCrl
- TriggerMbr
  remove unused TCI Messages
2. remove unresolved conflict message

cv2x commented on ASN1/TCIDispatcher.asn:
```diff
@@ -1,40 +1,43 @@
+    ...,
+    d16093cv2x [7] TCI16093PC5,
+    d31611     [8] TCI31611,
+    d160921    [9] TCI160921
```
Missing comma after the new value!.

cv2x reviewed (changes requested):
the "d160921 [9] TCI160921"  in the Dispatcher is missing a comma before the next value.

wooks-ssoltech committed:
fixed missing comma

cv2x commented on ASN1/TCIDispatcher.asn:
```diff
@@ -1,40 +1,43 @@
+    d16093cv2x [7] TCI16093PC5,
+    d31611     [8] TCI31611,
+    d160921    [9] TCI160921,
+    proxyCv2x  [16] TCIPROXYCV2X,
```
trailing "," here causes compilation issues in some tools.

cv2x reviewed (changes requested):
TCIPROXYCV2X has extra comma after it

wooks-ssoltech committed:
remove a extra comman after TCIPROXYCV2X

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,167 @@
+    eCcfDownResp            (6),
+    eCrlDownResp            (7),
+    eMbrTriggeredResp       (8),
+    exception               (15) --exception added with large value to allow for future expansion
```
add extension marker to the Dot2Dot1State Enum, to make it future/backwards compatible when new states added.

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,167 @@
+    exception               (15) --exception added with large value to allow for future expansion
+}
+
+-- SecurityResultCode ::= ENUMERATED {     
```
this seems Enum seems entirely commented out.. do ehtese go in the "description"    or meant to be used instead of the common /simple  "ResultCode" that is being used now?

wooks-ssoltech replied:
I am not the author of this parts.
I also think that “enumerated” has been changed to “description.”
@neil-keysight-com I think security error codes will be replaced by the UCAM. Could you answer the Sean's question?

neil-keysight-com replied:
Looking through the history of this PR, I see @wooks-ssoltech is the only author, and the above appeared as part of the commit “revised as Neil’s review” on July 19th. It is present in the ‘ssoltech / OmniAir-Test-Control-Interface’ repo, in commit ae3a1001890c47f514ce1b063fbeb189789c1bbc by @tetris02 

I agree with Seans thought that this could be an error and should be replacing the common ‘ResultCode’ for some options within the 1609.2.1 module. Although it may equally be for future use or a merge error.

cv2x commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,167 @@
+    securityResultCode      ResultCode,
+                            -- All errors are lumped into single result code
+                            -- Specific cause of error to be described in the SecResultParams.description    
+    description             OCTET STRING OPTIONAL,
```
what is format of this Octet String/valid values?  is this a list of the SecurityResultcodes, or should it be a VisibleString type intsead?

wooks-ssoltech replied:
It's also not my work, but In this context, VisibleString seems more appropriate.

Jeffzhaoal replied:
OCTET STRING is usually used for binary data information.
UTF8String with a reasonable size constraint might be a good fit.

wooks-ssoltech committed:
add extension marker in the Dot2Dot1State for the addtion of new state.

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+}
+
+SetInitialize ::= SEQUENCE {
+  bootstrapUrl     VisibleString,
```
IT seems unlikely that ust providing a URL can treigger bootstrapping.  Please add inline comments that explain what this URL is supposed to do.  Is changing this URL supposed to also trigger SUT to completely bootstrap for a new set of enrollment,   should SUT  aGenerating the EE's public/private key pair and identifier ?  
Should  SUT then also submit the public key and the canonical ID to the SCMS for enrollment registration and choose the appropriate SCMS policy and certificate profile for the EE?
If this is not implied, sugest adding a new message that does such things.

This seems like the only place where a FQDN might be supplied, but somehow for SUT to reach TS, it will require changing host-name to the TS host name/IP for not just the Initialization/bootstrap, but also supplying address for the other functions like:  the ECA & RA.  Suggest making it clear that host-names/IP are being supplied for all the TS server uses.

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+  ctlSeriesId      VisibleString,
+  ...
+}
+
```
how did the SUT get supplied over TCI wwith ECA URL for initial enrollment and RA URLs for subsequent OTA services.  Its not always identical tot the bootstrap URL

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+  ...
+}
+
+TriggerMbr ::= BOOLEAN
```
Need to make it clear.  is this supposed to trigger a connection to RA /MA  immediately? For which PSID'/RA's? (since SUT can have multiple in parallel)
Does it need to connect even if there are no MBR's to upload?

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+SetInitialize ::= SEQUENCE {
+  bootstrapUrl     VisibleString,
+  ...
+}
```
For bootstrapping, suggest we also make a TCI to retrieve Cannonical ID.

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+TCI-160921
+DEFINITIONS AUTOMATIC TAGS ::= 
+BEGIN
+
```
add comment block that explains the purpose of this module, and the functions it provides.  mention that it is meant to support OBU/RSU/AVD test cases and the OA-764b  TSS&TP

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+  ...
+}
+
+DownloadCcf ::= SEQUENCE {
```
Downloading CCF requires a CTLSeriesId for downloading a CCF that includes a CTL matching the given CTLSeries.  

Similarly, for downloading a composite CRL, the standard also requires a parameter of CTLSeriesId to download the composite CRL with a CTL. It appears that the two messages, DownloadCcf and DownloadCrl, are intended to download the most recent CCF and the composite CRL, including a CTL. 

If this is the case, plese specify in the comment on this command in the ASN.1 file. 

suggest for testing CCF and CRL test cases, we should consider the SeriesID

Jeffzhaoal replied:
Does DownloadCcf need a field of version? This could be helpful to match the SCMS REST API version or other use cases.

Jeffzhaoal replied:
Does the data needs a version field?

cv2x commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+  ...
+}
+
+DownloadCrl ::= SEQUENCE {
```
Speify in comment what this does?  is it “the most recent CRL with that cracaId and crlSeries known to the RA.”. 

“The most recent CRL” could be one of four choices: fullHashCrl, deltaHashCrl, fullLinkedCrl and deltaLinkedCrl. To make this command more efficient, is it possible to request:

1.	A composite CRL of given cracaId and crlSeries including its most recent full CRL and all subsequent delta CRLs
2.	A single secured CRL of given cracaId and crlSeries for a specific crlSerial for hash Id CRL
3.	A single secured CRL of given cracaId and crlSeries for a specific iRev and/or indexWithinI for linkage value CRL


can we test a devices ability to handle all types of CRL resposnme with this TCI?  Perhaps this is purely up to thje TS.

Jeffzhaoal replied:
Does DownloadCrl need a field of version? This could be helpful to match the SCMS REST API version or other use cases.

cv2x reviewed (changes requested):
mostly need for clarifying specification of the messages assumptiosn in comments, please. Without that internal docuemnation, its impossible to know if the TCI commands are sufficient

Jeffzhaoal commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+}
+
+DownloadCrl ::= SEQUENCE {
+  cracaId          VisibleString,
```
The IEEE 1609.2.1-2022 requires:

1. One parameter CTLSeriesId for downloading a composite CRL;
2. And two parameters cracaId and crlSeries to download an individual CRL.

Suggestions:
1). Adding a TCI command for downloading composite CRL which will ensure the DUI gets all up-to-date CRLs.
2). Clarifying the current "DownloadCcf" for command whether the DUT is expected to download most recent CRLs of all crlSeries associated with this cracaId; or, maybe the command can include the crlSeries as an optional parameter if it is suppose trigger downloading the individual CRL identified by the given cracaId and crlSeries.

Jeffzhaoal commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+
+SetInitialize ::= SEQUENCE {
+  bootstrapUrl     VisibleString,
+  ...
```
Does SetInitialize need a field of version? This could be helpful to match the SCMS REST API version or other use cases.

Jeffzhaoal replied:
After an initialization or bootstrapping (as indicated by the state diagram in 764b-OA-TSS&TP01609.2.1), does the TS need to verify that the DUT has installed correct information so it is ready for enrollment? For example, does after sending this command to the DUT, is the TS supposed to check the install trust anchor information, i.e., trusted elector hashes, or trusted CTL, etc.?

Jeffzhaoal commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+
+
+
+Request160921 ::=  SEQUENCE {
```
Does Request160921 need a field of version? This could be helpful to match the SCMS REST API version or other use cases.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+    eProvisioned            (5),
+    eCcfDownResp            (6),
+    eCrlDownResp            (7),
+    eMbrTriggeredResp       (8),
```
1. When a request is not satisfied, i.e., authorization certificate (AC) request is rejected, does the enum need a request rejected state?
2. Is "eProvisioned" intended for the DUT to indicate it has downloaded authorization certificates (ACs)and is able to validate them and use them?
3. Do the enum also need:
    a.  "CRL overdue" state in case the downloaded CRL is not up-to-date?
    b. "missing CA" when the downloaded CCF download include CA certificates for building a certificate chain for verification?
    c. enrollment certificate is blocked so the DUT cannot download ACs and request enrollment successor?

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduParsingUnsupportedCriticalInformationField (4),
+--    spduParsingCertificateNotFound (5),
+--    spduParsingGenerationTimeNotAvailable (6),
+--    spduParsingGenerationLocationNotAvailable (7),
```
For IEEE 1609.2 data, it is helpful to define spduParsingGenerationLocationNotAvailable.
According to IEEE 1609.2.1, Ieee1609Dot2Data-Signed's content.signedData.tbsData.headerInfo should not have generationLocation field, i.e. "generationLocation ABSENT", so this value doesn't seem to fit 1609.2.1  Ieee1609Dot2Data-Signed.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduConsistencyFutureCertificateAtGenerationTime (17),
+--    spduConsistencyExpiredCertificateAtGenerationTime (18),
+--    spduConsistencyExpiryDateTooEarly (19),
+--    spduConsistencyExpiryDateTooLate (20),
```
Can you clarify what spduConsistencyExpiryDateTooEarly (19) and spduConsistencyExpiryDateTooLate (20) are for?

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduConsistencyExpiryDateTooEarly (19),
+--    spduConsistencyExpiryDateTooLate (20),
+--    spduConsistencyGenerationLocationOutsideValidityRegion (21),
+--    spduConsistencyNoGenerationLocation (22),
```
According to IEEE 1609.2.1, Ieee1609Dot2Data-Signed's content.signedData.tbsData.headerInfo should not have generationLocation field, these two result code (21 and 22) don't seem to apply to IEEE 1609.2.1.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduConsistencyGenerationLocationOutsideValidityRegion (21),
+--    spduConsistencyNoGenerationLocation (22),
+--    spduConsistencyUnauthorizedPSID (23),
+--    spduInternalConsistencyExpiryTimeBeforeGenerationTime (24),
```
According to IEEE 1609.2.1, Ieee1609Dot2Data-Signed's content.signedData.tbsData.headerInfo should not have expiryTime field, the result code 24 doesn't seem to apply to IEEE 1609.2.1.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduLocalConsistencyChainWasTooLongForSDEE (29),
+--    spduRelevanceGenerationTimeTooFarInPast (30),
+--    spduRelevanceGenerationTimeTooFarInFuture (31),
+--    spduRelevanceExpiryTimeInPast (32),
```
Similar comment as result code 24.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduRelevanceGenerationTimeTooFarInPast (30),
+--    spduRelevanceGenerationTimeTooFarInFuture (31),
+--    spduRelevanceExpiryTimeInPast (32),
+--    spduRelevanceGenerationLocationTooDistant (33),
```
According to IEEE 1609.2.1, Ieee1609Dot2Data-Signed's content.signedData.tbsData.headerInfo, the result code 33 doesn't seem to apply to IEEE 1609.2.1.

Jeffzhaoal commented on ASN1/TCIindication.asn:
```diff
@@ -1,148 +1,168 @@
+--    spduCertificateChainInconsistentExpiryTimes (13),
+--    spduCertificateChainInconsistentStartTimes (14),
+--    spduCertificateChainInconsistentChainPermissions (15),
+--    spduCryptoVerificationFailure (16),
```
As the code 16 is for SPDU crypto verification failure, it is different from the crypto verification failure on a CA cert in the SPDU's signing certificate chain. A distinct result for the CA crypto verification failure seems needed.

Jeffzhaoal commented on ASN1/TCI160921.asn:
```diff
@@ -0,0 +1,89 @@
+}
+
+TriggerMbr ::= BOOLEAN
+
```
Suggestions for verification of Initialized&Unenrolled, Enrolled&UnAuthorized and Authorized states in 764b-OA:

If the TCI initialization request is for bootstrapping or preparing for OTA enrollment and certificate requests, it should typically occur early in a secure environment in reality. The device must download the latest CTL, CCF, and CRL through OTA as soon as it connects to the configured SCMS before enrollment. The OTA enrollment process requires generating the EC request and processing the EC response following security checks defined in 1609.2 and 1609.2.1, which may fail if the CTL, CCF, and CRL are not up-to-date. To ensure proper functionality, verify after initialization that the device:
- Automatically downloaded the latest CTL, CCF, and CRL through OTA services. (TS can check if the DUT has the latest files).
- Automatically requested EC, AC, and downloaded ACs for signing secure messages. (TS can check if the DUT has EC and ACs).

The DownloadCcf and DownloadCrl requests are not typically useful immediately after initialization when the DUT has network connectivity as these should be automatic. However, they can instruct the DUT to download specific files to set up conditions for other tests.
